### DISCUSSION TITLE
Elevation / speed diagram in route planning, GPS track recording, favorite routes / tracks

### DISCUSSION THREAD

**Discussion by devemux86** [devemux86/cruiser#411] [Q&A] [answered]
Graphs arrive in Cruiser. :chart_with_upwards_trend: 

Elevation graph is plotted on two axes: X-axis is the distance of the route / track and Y-axis is the elevation.

Gestures with the graph (similar to map):

- Zoom in / out. Use two-finger movement (or double tap) to zoom in / out for a more detailed view of elevation changes. Move the graph forward and backward on the route / track.

- Place a point on the graph. To see the elevation at a particular point on the route / track, tap anywhere on the graph and a pointer with the value will appear.

Elevation graph is available in:

- **Route planning**: see the planned route details.

- **GPS recording**: see the recorded track details.

- **My places**: see the favorite route / track details.

More route / track details may be added in the future.

<img src="https://github.com/devemux86/cruiser/assets/3484020/2e18d6fa-4341-4194-8d43-5638f5d10aab" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/77fdaed2-3194-4bc9-8999-e04654924b79" width="50%"/>
![Screenshot_1704959805](https://github.com/devemux86/cruiser/assets/3484020/3a44db88-89fd-4228-81c3-0f95fdf852e5)
![Screenshot_1704959775](https://github.com/devemux86/cruiser/assets/3484020/6df68191-6d8b-4f32-8d7d-65e3c47f2cba)
![Screenshot_1704959575](https://github.com/devemux86/cruiser/assets/3484020/c771bc7b-cc53-44a7-a3bd-071908aa5bd5)
![Screenshot_1704959601](https://github.com/devemux86/cruiser/assets/3484020/050b29ad-4934-4bce-8bb8-cde2c69e1437)

**Comment by HannesHoBu** [comment]
This looks great, @devemux86 will you integrate in this context the counting of curves?

**Reply by devemux86** [reply]
@HannesHoBu we need an algorithm to recognize the curves in the geometry of a path. :slightly_smiling_face: 

For reference, there is a relevant discussion here:

- https://github.com/devemux86/cruiser/discussions/342

**Comment by DolomitesTom** [comment]
cool... especially with speed data

**Comment by devemux86** [comment]
The first implementation is available for testing in:
- https://github.com/devemux86/cruiser/discussions/414

**Comment by devemux86** [comment]
[BRouter website](https://brouter.de/brouter-web/) has an interesting function where we can move the mouse over the graph
and a marker appears above the route showing the location of the data, like elevation.

We could probably add this feature to the Cruiser graphs as well, if there is interest.
(then graphs will not be full screen)

**Reply by umpfmpf** [reply]
Is it possible to display the graph in route planning inside the App? 
And I think it would be a good function to display the position on the map when clicking on the graph.

**Reply by devemux86** [reply]
@umpfmpf 

The graph is on one part of the screen while we do route planning on the rest of the screen?

It could be possible (that's also what I ask above). We just need to see how we mean it to work.

**Reply by umpfmpf** [reply]
Maybe it is possible to install an option to choose when the graph is displayed (Navigation, follow , allways).
The way it looks on the route screen is quite good!
Transparency (possibly adjustable) would still fit quite well. 
And if I move or zoom the axis, it would also be nice if a presetting was set again when centering by time or button. Perhaps also an own standard zoom Level for the graph.

**Reply by devemux86** [reply]
Follow location is a free mode that follows GPS location, regardless of whether there is a route / track or not.
While navigation processes the GPS location on the route, so the graph can recognize where we are moving.

The graph in navigation auto-centers after a while (if we scroll it).
(it uses the same location setting of the map auto center / zoom)

The graph currently does not auto zoom, so we can freely adjust its scale in navigation.
(but we could change this if necessary)

Graph navigation panel now uses the existing color / transparency settings.
If we need something separate we can implement it. :slightly_smiling_face:

**Reply by DolomitesTom** [reply]
Looks good. I would prefer the graph below any text, for example name of next waypoint, not above.

**Comment by bios4** [comment]
> Elevation graph is available in:
Route planning: see the planned route details.

Nice feature @devemux86!

Suggestions:
1) Add a "Route details" menu item to the route menu when long-pressing the route on the map; preferably right before the existing "Recalculate route" or "Delete route" entry.
2) Move the "Route details" menu button in "Menu > Routing" from its current position (after "Options" and before "Delete") to be displayed after "Turn instructions" and before "Options".

Are you planning to add more functionality to the graphs?

E.g. moving the map with a marker on the route, when selecting a certain point in the graph?
Or allowing to select a certain part of the route on the graph and displaying this particular selected route section then on the map?

EDIT: I just saw that you already wrote about such additional features above, sorry that I missed this in the first place...

**Reply by devemux86** [reply]
@bios4 thanks for checking all these details! :+1: 

(1), (2) are simple and will be available in the next update.

> Are you planning to add more functionality to the graphs?

This is just the first implementation, we will add many more functions.

**Comment by MotoUKRider** [comment]
Awkward user from the UK writes: We absolutely insist on making life difficult by operating our traffic signs in MPH and Yards. However, if you want to buy petrol - the price is per litre.  Vehicles refer to "miles per gallon".  Timber is sold "per metre". Beer is poured by the pint. We are a funny little island.

Is it possible, without upsetting too much behind-the-scenes, to allow for specific units of measurement for different "features"?  The graph and ASL would be better in Metres, but then I'd be stuck with KM for distances.  It would save me dividing the values by 3.28 in my head :smile: 

There must be other places that mix things like this, or at least people of a certain age who grew up with metric but have to obey old metal signs displaying fractions, ratios and percentages :confused:

**Reply by devemux86** [reply]
@MotoUKRider interesting topic. What specific items do you mean?

Should elevation always be in Metric system?
Shouldn't distance be in Imperial system?

**Reply by MotoUKRider** [reply]
@devemux86 Rather than a global "unit" setting, the ability to select our preference for whatever Distance/Height "thing" is being displayed.

**Reply by devemux86** [reply]
@MotoUKRider these are too many to add separate settings without some global rules.

About distance, we have it in route planning, voice navigation, graph, ...

For example, see how many combinations OsmAnd has:
<img src="https://github.com/devemux86/cruiser/assets/3484020/17986a59-93b6-4f7c-8d96-10593547cffb" width="50%"/>

**Reply by MotoUKRider** [reply]
@devemux86 A setting offering those types of combinations would work.  "Miles/meters" would be my ideal.
My main concern was how easy it was to calculate inside the code before working out the GUI aspect for us mortals :confused:

**Reply by devemux86** [reply]
@MotoUKRider I will put it in my list to check what is required.

It needs to split the unit settings per type (distance, speed, etc).

**Comment by devemux86** [comment]
How should we handle the route graph outside of navigation?

- Add option to make it available in "Browse map" so we can have it always visible in route planning.
  - Then we can move its option in "Application | Display" settings.

- We could also show the graph in "Follow location" if there is a route?
No line progress as we follow the GPS regardless if there is a route or track.

**Comment by devemux86** [comment]
GPS recording also stores the speed in addition to the elevation. So its graph can display both values:

![Screenshot_1704959575](https://github.com/devemux86/cruiser/assets/3484020/c771bc7b-cc53-44a7-a3bd-071908aa5bd5)
![Screenshot_1704959601](https://github.com/devemux86/cruiser/assets/3484020/050b29ad-4934-4bce-8bb8-cde2c69e1437)

**Reply by devemux86** [reply]
We will also have the option of displaying Elevation or Speed or both values on the graph.

<img src="https://github.com/devemux86/cruiser/assets/3484020/2c093577-a9db-4be7-8dd6-3a1a780f78ef" width="50%"/>

**Reply by DolomitesTom** [reply]
perfect :-)

**Reply by umpfmpf** [reply]
looks great!

**Reply by devemux86** [reply]
The color of the 1st value (elevation) is the app accent color (selected by user).
This is nice as the graphs can have any color we choose.

But when we display 2 values (elevation + speed), it is difficult to decide a 2nd good color when the 1st color is not fixed.
Unless we use fixed colors when we display multiple values, e.g. red + blue?

**Reply by umpfmpf** [reply]
Would it be possible to place a selection of 4-5 fixed colors next to or under the selection fields for elevation and speed?

**Reply by devemux86** [reply]
@umpfmpf we will improve the charts more in the future with each new update.

For a start we just need to have the basics ready so that they can be easily used.

**Reply by bios4** [reply]
> The color of the 1st value (elevation) is the app accent color (selected by user).
But when we display 2 values (elevation + speed), it is difficult to decide a 2nd good color when the 1st color is not fixed.

Is it programmatically possible to simply determine the "complementary color" of the user-selected app color and apply it to the second chart?

**Reply by devemux86** [reply]
Please see the updated images [above](https://github.com/devemux86/cruiser/discussions/411#discussioncomment-8067219).

It is not just the 2 colors for the data, they also need to be visible on the background (light or dark).
And if we think the people with color problems, the automatic colors are more difficult to decide.

**Comment by devemux86** [comment]
Waypoints will be visible in the route graphs:

![Screenshot_1704959805](https://github.com/devemux86/cruiser/assets/3484020/3a44db88-89fd-4228-81c3-0f95fdf852e5)
![Screenshot_1704959775](https://github.com/devemux86/cruiser/assets/3484020/6df68191-6d8b-4f32-8d7d-65e3c47f2cba)

**Comment by devemux86** [comment]
> Is it possible to display the graph in route planning inside the App?
And I think it would be a good function to display the position on the map when clicking on the graph.

It will be available in the next Cruiser update. :slightly_smiling_face:

**Comment by devemux86** [comment]
Route planning with graph:
- Tap or drag on the graph to move the map to the route position
- Zoom in / out the graph to see more route details
- Turn on / off graph by long press on route or in routing settings

<img src="https://github.com/devemux86/cruiser/assets/3484020/2e18d6fa-4341-4194-8d43-5638f5d10aab" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/77fdaed2-3194-4bc9-8999-e04654924b79" width="50%"/>

**Reply by bios4** [reply]
> Tap or drag on the graph to move the map to the route position

Just installed b13...

Example navigation is started and route diagram is displayed.
Tapping on the graph does "nothing".

So how exactly is this supposed to work???

> Zoom in / out the graph to see more route details

Zooming in the graph does not change my map view/displayed map area, but only the graph.
Is this supposed to work like this or what should happen when zooming?


> Turn on / off graph by long press on route or in routing settings

This option is only when navigation is not started (yet).
Could this maybe be changed to allow de-/activating the graph also during a running navigation?

**Reply by devemux86** [reply]
@bios4 this is to display a graph in route planning with its own on / off option.

Navigation is a different process with its own graph, displaying the progress.

There are separate on / off options depending on what we want to use each time.

**Reply by bios4** [reply]
Oh I overlooked that this ^^ is about the planning part of Cruiser.
Sorry...

**Comment by Ciutnik** [comment]
Looks good. Probably one day we will be able to show any touched point of the graph on the map? 🙂

The only thing I don't like is the menu name "Route details". Subjectively, I associate it more with some numerical data. Why not a Route chart or graph? The same applies to the track.

**Reply by devemux86** [reply]
@Ciutnik we could display more data in the future in this screen, not just graph.
And already distance / time / ascend / descend are displayed besides the graph.

**Reply by devemux86** [reply]
However, we could do rename them, now only 11 languages. :slightly_smiling_face:

**Reply by bios4** [reply]
> Probably one day we will be able to show any touched point of the graph on the map?

Use case 1:
During route planning open the height diagram/route graph via Menu > Routing > Route details; when tapping/dragging on the graph currently a "crosshair-styled" pointer displays the height/distance of the respectve point of the route. Wish 1: when closing the diagram, it would be extremely nice when the map would be centered on this exact point I marked in the height diagram/route graph.

Use case 2:
During route planning the height diagram/route graph is displayed on the top of the screen; when tapping/dragging on the graph the map currently already moves and shows the respective point I selected in the graph in the display center; unfortunately without any indicator... Wish 2: a crosshair in this very display center, indicating the exact point on the route I selected in the height diagram/route graph.

What do you think @devemux86?

**Reply by Ciutnik** [reply]
Maybe a map button in the chart window, known from another application?

<img src="https://github.com/devemux86/cruiser/assets/122647782/0f626894-e8cd-4c97-a495-7a2ebfe8ab79" width="50%"/>

**Reply by devemux86** [reply]
1. This is easy, see the image.

3. This needs some thought for the implementation.

![Screenshot_1705758396](https://github.com/devemux86/cruiser/assets/3484020/99eb52e7-26a3-440b-9f2f-d09ca2c16ed3)

**Reply by devemux86** [reply]
2. Next update will display a (non-permanent) crosshair at the route location (map center) when the route diagram is clicked.

**Comment by Ciutnik** [comment]
I can't recognize what the `Elevation graph` switch in `Settings | Routing` is for. We have the same one in `Settings | Navigation | Display`, but here the matter is obvious.

**Reply by devemux86** [reply]
These are 2 completely different modes:
- Navigation
- Route planning

<img src="https://github.com/devemux86/cruiser/assets/3484020/75df8ff5-2b6d-477f-9667-151bbdd508ac" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/2e18d6fa-4341-4194-8d43-5638f5d10aab" width="50%"/>

**Reply by Ciutnik** [reply]
Thanks. I see already! 👍

In route planning mode, I had the route created in older beta (it was already there before I updated the app) and therefore couldn't see the chart 🙂

Of course, when creating a NEW route, the graph appears as expected.

**Reply by devemux86** [reply]
The diagram appears on screen or is offered in menu when there is the data in the route or track, like the elevation.

We could have routes or tracks without elevation. In the future we can display more data types.

**Comment by DolomitesTom** [comment]
Would it be possible if you zoom in the map and move while planning that the graph bar is also positioning in this area? Maybe take the center of the map as marker position for the graph bar?

**Reply by devemux86** [reply]
@DolomitesTom the diagram on the X-axis has the route distance, not map coordinates.

So when we move the map, we have to translate it to somehow "travel" on the route to make this possible?
Something like what happens in navigation, we move on the route, the location bar moves on the diagram.

**Reply by DolomitesTom** [reply]
Yes something like this if it is not too complicated. I am more interested in some statistics in the track details like average speed with and without pauses, driven kilometers, overall time with and without pauses including storing these values in the exported files or track menu and restoring these values when loading tracks again.

**Comment by devemux86** [comment, marked as answer]
Elevation / speed diagram in route planning, GPS track recording, favorite routes / tracks is available in [Cruiser 3.0.38](https://play.google.com/store/apps/details?id=gr.talent.cruiser).
